GoogleForCreators/web-stories-wp
# Text Style: Changing font family should result nearest font weight

_(Updated 6/12/2020)_

## Summary
- When there is only 1 font weight in an entire text element and you switch font families, you will not end up with "mixed" - formerly "multiple" (placeholder was updated over the lifespan of this branch). 
- When an existing font weight is not available when you make said switch of font families, we will get the closest available weight to that previously existing weight. 
- When you have multiple font weights and you switch weights, if you have the entire text element in focus you will still see "mixed", BUT when you enter that text element, if you had "light" selected but now only regular or bold are options, it'll update the selected weight to be regular when you make that portion of text element active again. 


## Relevant Technical Choices
- This has been tricky! I added notes inline to the useEffect hook that's handling these updates. I think walking through that is the easiest way to note what happened. Here's the TL;DR

- new util called `getClosestFontWeight` that will return the weight value closest to the desired value should that desired value be unavailable. 
- Created refs for both fontWeight and fontFamily because we get a flash of default fontWeight and family that flashes bad values for the weight dropdown as well as interrupting our check of if the set font weight is available. By having refs we can check and see if what we are looking at is current or not based on if the font family or weight changed actively by the user. 
- Added a `useEffect` hook that does the following in this order (returns if a criteria is met): 
```
1. If the `fontFamily` doesn't match the `fontFamilyRef` then we ignore any potential weight changes. `fontFamilyRef` stays true to the `fontFamily` chosen by the user regardless of default content or flashes back to reset depending on where you are clicking or selecting. 
2. if there is only 1 available weight in the given set of `fontWeights` then we know that's the weight we want. Check to see if this weight was already updated to (this check only needs to be there for when the font family or font weight select menu is expanded again with that default value sneaking through). If the available weight doesn't match the ref weight then we update the font weight and ref. 
3. if the `fontWeight`'s true value is "multiple" then we just leave it at this point. We know there are more than one available weight values, that's enough. 
4. If we have gotten this far it means we are inline to the text element. So we check to see if the `fontWeight` that was/is selected is available to our list of weights and then we check to see if that value is the same as the existing value. If they are different, we update the value and the ref with the new value. 
5. Lastly, if we got here we can trust the content and should just keep the ref current. 
```

- Anddd I memoized the visible value of the font weight on select menu. This was getting a little complicated and I didn't want to add a second ternary to do a check on how many available font weights were present. Now we are doing that same check for flashing default font family/return back to the old font family value when updating it, and then within the "multiple value" we're making sure that there are more than 1 option for value, when it's just the one i'm forcing that as the value. 


- This bug is tricky because we can't possibly know where in the text element is active. It could be the entire element, a highlight of text, or just the cursor active in the field. I tried really hard to stay contained in this specific new piece of functionality to avoid a trickle of potential new bugs stemming from altering functionality of how anything is focused. I couldn't just update the font weight for when "mixed" was set without having the text element blow up on me when i entered into it occassionally. I've found this approach to be more congruent, if more involved. 

- I did make one update outside of textStyle/font that I want to draw attention to - `richText/useSelectionManipulation`. I was noticing that the text elements were trying to update with old/null data sometimes if you click too fast on an element. I added an early return for if the `oldState` doesn't exist in the `updateWhileUnfocused` check, this prevents the editor error-ing out for me when I'm just waiting for data to return. 


## To-do

- I'm hoping that my novel above helps point out a better way to handle this without needing early escapes for out of date data in the future
- Also, if this is an appropriate solution then I have some unit tests to add and update. 
- Sometimes the font weight that's shown as the value in the dropdown doesn't update from "mixed" - I've found that this stems from swapping a font family, then adjusting inline weight and then changing the font family again without clearly updating the focus of the text element to the whole thing. If anyone else can point out how to manage this without having other issues pop up I'm all ears, for me, this is an OK edge case for the sake of the main fix. it seems like something to do with the currently selected item not being current - last week i traced this towards the fauxSelection process but I don't know enough about how that whole process works and this feels like a separate issue.
- Also of note, if you have mixed font weights and swap fonts to a font that doesn't support 1 of the weights but it still has some weight that was set inline, if you swap back to a font family that does support that existing inline weight, it'll still exist.I don't really think there's a way to fix that here within the existing structure of how focus is obtained and maintained. It's one of the reasons why this has gotten so tricky.

## User-facing changes
- See summary


## Testing Instructions

1. If a text element has multiple weights applied to it, when a new font family is picked the weight for the entire text element should  remain "mixed" if you have the entire text element in focus. 
2. if a text element has multiple weights applied to it, but you are INLINE of the text element when a new font family is picked then the shown font weight value in the dropdown will be the font weight that it was in the previous font family selected OR if that weight is no longer available it will be the closest weight to that old value. 
3. If you have a nice, simple text element that only has one weight and you update the font family, if that font weight is not available in the new family's weights then the closest weight to the weight that was selected will be applied and visible. aka no more 'mixed'. The caveat here is that note in the "to do" section, where sometimes the set selection focus is out of sorts with what you're trying to update and you might see Mixed. 


Bug reproduction: 
```
Add text element
Assign font family Open Sans Condensed, font weight Light
Now change the font family to Playball
See "multiple" as the font weight.
```

Now:
```
Add text element
Assign font family Open Sans Condensed, font weight Light
Now change the font family to Playball
**See "Regular" as the font weight.**
```



---



Addresses #1899
